Installation issue: gcc+nvptx

### Steps to reproduce the issue

There seems to be a problem building gcc with nvptx support due to a recent change in the CUDA package.

<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->

```console
$ spack install gcc@9.3.0+nvptx
...
 >> 34532    configure: error: CUDA driver package required for nvptx support
..
 >> 34535    make[1]: *** [configure-target-libgomp] Error 1
```

### Information on your system

$ spack debug report
* **Spack:** 0.15.4
* **Python:** 3.6.8
* **Platform:** linux-centos7-skylake_avx512

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

### Additional information

This looks to be due to the following commit which excludes the stubs directory from the CUDA library path:

https://github.com/spack/spack/commit/54bce00d4dd606b20f8b6ad1efad58d2fc0fddeb

I suspect that configure is looking for libcuda.so which is not included in the runtime libraries (it comes with the driver) but there is a stub present.

I can build correctly using the previous version of the CUDA package.py 

@michaelkuhn - I'm not sure if this should be considered a GCC issue or a CUDA one but the commit by @haampie mentioned above isn't daft as far as CUDA itself goes. 

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [ ] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -128,7 +128,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('gnat', when='languages=ada')
     depends_on('binutils+ld+plugins~libiberty', when='+binutils', type=('build', 'link', 'run'))
     depends_on('zip', type='build', when='languages=java')
-    depends_on('cuda@:10', when='+nvptx')
+    depends_on('cuda', when='+nvptx')
 
     # The server is sometimes a bit slow to respond
     timeout = {'timeout': 60}
@@ -145,7 +145,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     # nvptx-tools doesn't have any releases, so grabbing the last commit
     resource(name='nvptx-tools',
              git='https://github.com/MentorEmbedded/nvptx-tools',
-             commit='5f6f343a302d620b0868edab376c00b15741e39e',
+             commit='d0524fbdc86dfca068db5a21cc78ac255b335be5',
              when='+nvptx')
 
     # TODO: integrate these libraries.
@@ -441,6 +441,19 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
             filter_file('@zlibinc@',
                         '-I{0}'.format(spec['zlib'].prefix.include),
                         'gcc/Makefile.in')
+
+        if spec.satisfies('+nvptx'):
+            # backport of 383400a6078d upstream to allow support of cuda@11:
+            filter_file('#define ASM_SPEC "%{misa=*:-m %*}"',
+                        '#define ASM_SPEC "%{misa=*:-m %*; :-m sm_35}"',
+                        'gcc/config/nvptx/nvptx.h',
+                        string=True)
+            filter_file('Target RejectNegative ToLower Joined '
+                        'Enum(ptx_isa) Var(ptx_isa_option) Init(PTX_ISA_SM30)',
+                        'Target RejectNegative ToLower Joined '
+                        'Enum(ptx_isa) Var(ptx_isa_option) Init(PTX_ISA_SM35)',
+                        'gcc/config/nvptx/nvptx.opt',
+                        string=True)
 
     # https://gcc.gnu.org/install/configure.html
     def configure_args(self):

--- a/var/spack/repos/builtin/packages/nvptx-tools/package.py
+++ b/var/spack/repos/builtin/packages/nvptx-tools/package.py
@@ -15,6 +15,7 @@ class NvptxTools(AutotoolsPackage):
     homepage = "https://github.com/MentorEmbedded/nvptx-tools"
     git      = "https://github.com/MentorEmbedded/nvptx-tools"
 
+    version('2021-05-21', commit='d0524fbdc86dfca068db5a21cc78ac255b335be5')
     version('2018-03-01', commit='5f6f343a302d620b0868edab376c00b15741e39e')
 
     depends_on('binutils')

